pitop/processing/algorithms/ball_detect.py

```diff
@@ -0,0 +1,247 @@
+
+        if colour.lower() not in ('red', 'green', 'blue'):
+            raise ValueError('Colour must be "red", "green" or "blue"')
+
+        blurred = cv2.blur(frame, (11, 11))
+        hsv = cv2.cvtColor(blurred, cv2.COLOR_BGR2HSV)
+
+        masks = []
+        for colour_range in colour_ranges[colour]:
+            hsv_lower = colour_range['lower']
+            hsv_upper = colour_range['upper']
+            mask = cv2.inRange(hsv, hsv_lower, hsv_upper)
+            masks.append(mask)
+        mask = sum(masks)
+
+        mask = cv2.erode(mask, None, iterations=1)
+        mask = cv2.dilate(mask, None, iterations=1)
+
+        if not return_binary_mask:
+            mask = cv2.bitwise_and(frame, frame, mask=mask)
+
+        if output_format.lower() == 'pil':
+            mask = ImageFunctions.convert(mask, format="PIL")
+
+        return mask
+
+    @staticmethod
+    def __find_most_likely_ball(contours, colour, resized_frame):
+        ball_center = None
+        ball_radius = None
+        if len(contours) > 0:
+
+            max_likelihood_index = 0
+            for contour in contours:
+                # loop through all contours and store the most likely one's parameters
+                (x, y), radius = cv2.minEnclosingCircle(contour)
+                area = cv2.contourArea(contour)
+
+                # compare found contour with a perfect circle contour to dismiss coloured objects that aren't round
+                mask_to_compare = np.zeros(resized_frame.shape[:2], dtype="uint8")
+                cv2.circle(mask_to_compare, (int(x), int(y)), int(radius), 255, -1)
+                contours_compare = cv2.findContours(mask_to_compare, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+                contours_compare = grab_contours(contours_compare)
+                c2 = max(contours_compare, key=cv2.contourArea)
+                match_value = cv2.matchShapes(contour, c2, 1, 0.0)  # closer to zero is a better match
+
+                # most likely ball is a mixture of largest area and one that is most "ball-shaped"
+                likelihood_index = area / (match_value + 1e-5)
+
+                # check if this contour is more likely than the last
+                if likelihood_index > max_likelihood_index:
+                    # save to new max
+                    max_likelihood_index = likelihood_index
+                    # check if the ball is circular. Or if it's so big it's just close to the camera and occluded
+                    if match_value < ball_match_limits[colour] or radius > BALL_CLOSE_RADIUS:
+                        # check if it meets the minimum ball radius
+                        if radius > MIN_BALL_RADIUS:
+                            # ok, we're finally happy that this is an actual ball
+                            # store the values to be used as the most likely ball in frame
+                            ball_center = (int(x), int(y))
+                            ball_radius = int(radius)
```

Comment:
Too many comments - would like to see more clarity provided by breaking up functions and providing good names for the discrete steps so that - at a glance - it is more obvious what is going on.

Reply:
yeah agree, I mentioned I wasn't happy with this function to jorge but not sure how to improve it

Reply:
Refactored, let me know what you think